modules/apps/frontend-js/frontend-js-importmaps-extender/src/main/java/com/liferay/frontend/js/importmaps/extender/internal/servlet/taglib/JsImportmapsExtenderTopHeadDynamicInclude.java

```diff
@@ -0,0 +1,178 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.frontend.js.importmaps.extender.internal.servlet.taglib;
+
+import com.liferay.portal.kernel.json.JSONFactory;
+import com.liferay.portal.kernel.json.JSONObject;
+import com.liferay.portal.kernel.servlet.taglib.BaseDynamicInclude;
+import com.liferay.portal.kernel.servlet.taglib.DynamicInclude;
+import com.liferay.portal.url.builder.AbsolutePortalURLBuilder;
+import com.liferay.portal.url.builder.AbsolutePortalURLBuilderFactory;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Iván Zaera Avellón
+ */
+@Component(
+	immediate = true,
+	service = {
+		DynamicInclude.class, JsImportmapsExtenderTopHeadDynamicInclude.class
+	}
+)
+public class JsImportmapsExtenderTopHeadDynamicInclude
+	extends BaseDynamicInclude {
+
+	@Override
+	public void include(
+			HttpServletRequest httpServletRequest,
+			HttpServletResponse httpServletResponse, String key)
+		throws IOException {
+
+		PrintWriter printWriter = httpServletResponse.getWriter();
+
+		printWriter.println("<script type=\"importmap\">");
+
+		printWriter.println(_importmaps.get());
+
+		printWriter.println("</script>");
+
+		printWriter.print("<script src=\"");
+
+		AbsolutePortalURLBuilder absolutePortalURLBuilder =
+			_absolutePortalURLBuilderFactory.getAbsolutePortalURLBuilder(
+				httpServletRequest);
+
+		printWriter.print(
+			absolutePortalURLBuilder.forModuleScript(
+				_bundleContext.getBundle(),
+				"/es-module-shims/es-module-shims.js"
+			).build());
+
+		printWriter.println("\"></script>\n");
+	}
```

Comment:
So will there just be one importmap per page? Or would we ever want multiple on the page?

Reply:
There's only one importmap per page by design and it must be before any `<script type="module">` tag, otherwise it fails (this is logical, since the importmap changes the resolution algorithm, which could render the resolution in previously appearing modules invalid).

What we do is to allow contributions to the global importmap so that it can be modularized internally.